### PR TITLE
rust-overlay: install bash_completions to share/

### LINKED ISSUE #37: rust-overlay: nightly 2017-05-19 tries to write to /etc
The build fails with `mkdir: cannot create directory '/etc/bash_completion.d': Permission denied`

### PR DESCRIPTION
**Pull request by steveeJ**
Tested with `rustc 1.19.0-nightly (d47cf08d5 2017-05-28)`

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/nixpkgs-mozilla/38)

### CODE REVIEW

**Comment by zimbatm** [comment]
Do you know why install.sh doesn't install the man pages already?

**Comment by steveeJ** [comment]
> Do you know why install.sh doesn't install the man pages already?

I don't know why I had put the manpage example here, probably because I was testing availability of man pages and bash completions at the same time. However, it showed that they are installed as expected to _$out/share/man/_.

**Comment by zimbatm** [comment]
Ok I was confused.

Did you try setting `$CFG_SYSCONFDIR` to `$out/share`? I think that might solve the problem in a more succinct way. See https://github.com/rust-lang/rust-installer/blob/655a137a8bd0dda1cc71fb82661446e215563d6f/install-template.sh#L621

File: rust-overlay.nix [hunk 1]
```diff
@@ -105,9 +105,48 @@ let
...
+                  local wanted_path="$2"
+
+                  # Rename the files
+                  if [ -d "./$subdir/$orig_path" ]; then
+                    mkdir -p "$(dirname ./"$subdir"/"$wanted_path")"
+                    mv -v "./"$subdir"/"$orig_path"" "./"$subdir"/"$wanted_path""
+                  fi
+
+                  # Replace all occurences in the manifest file
+                  sed -i -e "s,"$orig_path","$wanted_path"," "$subdir"/manifest.in
```
Comment (nbp) [inline rust-overlay.nix]: nit: Use `SubstituteInPlace` [1] instead, which is the recommended way in Nixpkgs.

[1] https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L412

File: rust-overlay.nix [hunk 2]
```diff
@@ -105,9 +105,48 @@ let
             # (@nbp) TODO: Check on Windows and Mac.
             # This code is inspired by patchelf/setup-hook.sh to iterate over all binaries.
             installPhase = ''
+              handleEtc() {
+                local subdir="$1"
+                local oldIFS="$IFS"
+
+                # Directories we are aware of
+                for paths in \
+                  "etc/bash_completion.d","share/bash_completion/completions";
```
Comment (nbp) [inline rust-overlay.nix]: nit: Add a comment that these are substitution lists.

File: rust-overlay.nix [hunk 3]
```diff
@@ -105,9 +105,48 @@ let
             # (@nbp) TODO: Check on Windows and Mac.
             # This code is inspired by patchelf/setup-hook.sh to iterate over all binaries.
```
Comment (nbp) [inline rust-overlay.nix]: nit: This comment is about the `setInterpreter` function, move it above the function.

**Comment by nbp** [comment]
@zimbatm I agree that a more succinct way would be ideal, in the mean time if this is working I am ok with this changes.

The fact that `$CFG_SYSCONFDIR` is not set properly is also a problem that should be reported upstream if not done yet.

**Comment by steveeJ** [comment]
@zimbatm 
> Did you try setting $CFG_SYSCONFDIR to $out/share? I think that might solve the problem in a more succinct way. See https://github.com/rust-lang/rust-installer/blob/655a137a8bd0dda1cc71fb82661446e215563d6f/install-template.sh#L621

Since https://github.com/rust-lang/rust-installer/compare/daa2a05ebe7b8d07a309e8891ebc548652362954...655a137a8bd0dda1cc71fb82661446e215563d6f#diff-5c4807413996d3021c775ec958c4488f,  `$CFG_SYSCONFDIR` lives under `--prefix` which we set to `$out`.

@nbp 
> The fact that $CFG_SYSCONFDIR is not set properly is also a problem that should be reported upstream if not done yet.

Thanks for hinting at the upstream fix. I've reduced the PR accordingly. This won't help anyone who wants to use the interim, but this nightly and it will break every now and then ;-) 

I wanted to keep the history for discussions. Let me know if I should squash.

**Comment by nbp** [comment]
Sorry for the delay, I was a little slow to merge this change because it sounded like a lot of code for just a simple file.  And, then I forgot about this changes.

So better merged, than forgotten.  Then we can simplify this code later.
Thanks for investigating and fixing this issue.

**Pull request merged**